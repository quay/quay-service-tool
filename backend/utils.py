from flask_restful import Resource
from flask import request, make_response
from flask import current_app as app
from flask_login import current_user
from datetime import datetime
import logging
import json

logger = logging.getLogger(__name__)


def is_valid_severity(severity):
    return severity in ["default", "success", "info", "danger", "warning"]


class User(object):
    def __init__(self, is_authenticated=False, email=None, username=None, realm_access=None):
        self.email = email
        self.username = username
        self.is_authenticated = is_authenticated
        self.realm_access = realm_access


class Auth(Resource):
    def authenticate_user(token_info, config):
        if not token_info or not token_info.get("realm_access"):
            return User()

        realm_access = token_info["realm_access"]
        if not realm_access.get("roles") or not config["roles"] or \
           not (set(realm_access["roles"]) & set(config["roles"].values())):
            return User()

        return User(
            email=token_info["email"],
            username=token_info["name"],
            is_authenticated=True,
            realm_access=realm_access,
        )


def create_transaction(db):
    if type(db.obj).__name__ == "ObservableRetryingMySQLDatabase":
        try:
            db.close()
        except:
            pass

    return db.transaction()


def log_response(func):
    def wrapper(*args, **kwargs):
        response = func(*args, **kwargs)
        log_fn = logging.error
        if response.status_code == 500:
            log_fn = logging.exception
        if response.status_code == 200 or response.status_code == 201:
            log_fn = logging.info
        log_fn(
            f"{datetime.utcnow().strftime('%d %b, %Y, %H:%M:%S')} - "
            f"{request.method} - "
            f"{request.path} - "
            f"{response.status_code} - "
            f"{current_user.username} - "
            f"{current_user.email} - "
            f"{json.dumps(request.json)} - "
            f"{response.data}"
        )
        return response

    return wrapper


def verify_admin_permissions(func):
    def wrapper(*args, **kwargs):
        # Bypassing checks for local dev when auth is not on
        if app.config.get('is_local') and not app.config.get('test_auth'):
            return func(*args, **kwargs)

        if not current_user or not current_user.realm_access:
            return make_response(json.dumps({"message": "No RBAC defined for user"}), 401)

        ADMIN_ROLE = app.config.get('authentication', {}).get('roles', {}).get('ADMIN_ROLE')
        if ADMIN_ROLE not in current_user.realm_access.get('roles', None):
            return make_response(json.dumps({"message": "Require ADMIN permissions to perform action"}), 401)

        response = func(*args, **kwargs)
        return response
    return wrapper


def verify_export_compliance_permissions(func):
    def wrapper(*args, **kwargs):
        # Bypassing checks for local dev when auth is not on
        if app.config.get('is_local') and not app.config.get('test_auth'):
            return func(*args, **kwargs)

        if not current_user or not current_user.realm_access:
            return make_response(json.dumps({"message": "No RBAC defined for user"}), 401)

        EXPORT_COMPLIANCE_ROLE = app.config.get('authentication', {}).get('roles', {}).get('EXPORT_COMPLIANCE_ROLE')
        if EXPORT_COMPLIANCE_ROLE not in current_user.realm_access.get('roles', None):
            return make_response(json.dumps({"message": "Require EXPORT_COMPLIANCE_ROLE permissions to perform action"}), 401)

        response = func(*args, **kwargs)
        return response
    return wrapper


def verify_admin_or_export_perm(func):
    def wrapper(*args, **kwargs):
        # Bypassing checks for local dev when auth is not on
        if app.config.get('is_local') and not app.config.get('test_auth'):
            return func(*args, **kwargs)

        if not current_user or not current_user.realm_access:
            return make_response(json.dumps({"message": "No RBAC defined for user"}), 401)

        ADMIN_ROLE = app.config.get('authentication', {}).get('roles', {}).get('ADMIN_ROLE')
        EXPORT_COMPLIANCE_ROLE = app.config.get('authentication', {}).get('roles', {}).get('EXPORT_COMPLIANCE_ROLE')
        if not (set([ADMIN_ROLE, EXPORT_COMPLIANCE_ROLE]) & set(current_user.realm_access.get('roles', None))):
            return make_response(json.dumps({"message": f"Not enough permissions to perform action"}), 401)

        response = func(*args, **kwargs)
        return response
    return wrapper
