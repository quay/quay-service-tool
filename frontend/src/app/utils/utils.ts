export function accessibleRouteChangeHandler() {
  return window.setTimeout(() => {
    const mainContainer = document.getElementById('primary-app-container');
    if (mainContainer) {
      mainContainer.focus();
    }
  }, 50);
}

export function getErrorMessage(error){
  let message = "";
  if (error.response) {
    message = `Recieved HTTP status ${error.response.status} from server. Response: ${error.response.data.message}`;
  } else if (error.request) {
    message = "Unable to get response from server";
  } else {
    message = "Error initializing request";
  }
  return message
}
