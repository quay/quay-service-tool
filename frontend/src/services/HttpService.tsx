import axios from "axios";
import UserService from "./UserService";

const HttpMethods = {
  GET: 'GET',
  POST: 'POST',
  DELETE: 'DELETE',
};

const _axios = axios.create();

const configure = () => {
  _axios.interceptors.request.use((config) => {
    if (process.env.NODE_ENV !== "production") {
      return config;
    }
    else if (UserService.isLoggedIn()) {
      const cb = () => {
        config.headers.Authorization = `Bearer ${UserService.getToken()}`;
        return Promise.resolve(config);
      };
      return UserService.updateToken(cb);
    }
  },function(error) {
    return Promise.reject(error);
  });
};

const axiosClient = _axios;

const HttpService = {
  HttpMethods,
  configure,
  axiosClient,
};

export default HttpService;
