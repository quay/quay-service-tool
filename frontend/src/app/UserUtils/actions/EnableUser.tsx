import * as React from 'react';
import { useState } from 'react';
import {
    ActionGroup,
    Alert,
    AlertActionCloseButton,
    Button,
    Card,
    CardBody,
    CardTitle,
    Form,
    FormGroup,
    Modal,
    ModalVariant,
    PageSection,
    TextInput,
  } from '@patternfly/react-core';
import HttpService from "../../../services/HttpService";
import { getErrorMessage } from '@app/utils/utils';

type Props = {

};

type Alert = {
    variant: "success" | "danger" | "default" | "warning" | "info" | undefined;
    show: boolean;
    title: string;
}

export const EnableUser: React.FunctionComponent = (props: Props) => {
    const [username, setUsername] = useState('');
    const [alert, setAlert] = useState<Alert>({
        variant: 'success',
        show: false,
        title: ''
    });
    const [showModal, setShowModal] = useState(false);

    async function promptEnableUser(){
        if(username.length == 0){
            setAlert({ variant: 'danger', show: true, title: 'Please enter a username' });
            return;
        }
        HttpService.axiosClient.get(`/user/${username}`)
            .then(function(response){
                if(response.data.enabled){
                    setAlert({variant: "danger", show: true, title: `User ${username} is already enabled`});
                } else {
                    setShowModal(true);
                }
            })
            .catch((error) => {                 
                let errMessage = error.response.status == 404 ? `User ${username} does not exist` : getErrorMessage(error);
                setAlert({variant: "danger", show: true, title: errMessage})  
            });
    }

    async function enableUser(){
        HttpService.axiosClient.put(`/user/${username}?enable=true`)
          .then(function(response){
            setShowModal(false);
            setAlert({variant: "success", show: true, title: `User ${response.data.user} enabled`});
          })
          .catch((error) => { 
            setShowModal(false);
            setAlert({variant: "danger", show: true, title: getErrorMessage(error)}) 
        });
    }
    
    let modal: JSX.Element | null = null;
    if(showModal){
        modal = (
            <Modal
            title={`Enable user ${username}?`}
            isOpen={showModal}
            variant={ModalVariant.small}
            aria-label="feedback modal"
            showClose={true}
            onClose={() => {setShowModal(false)}}
            aria-describedby="no-header-example"
            >
                <Button id="enable-user-confirm" variant="primary" onClick={enableUser}>Enable User</Button>
            </Modal>
        );
    }
    return (<>
        <PageSection>
            <Card>
                {modal}
                <CardTitle className={"text-uppercase"}> Enable User </CardTitle>
                {alert.show && <Alert id="enable-user-alert" isInline actionClose={<AlertActionCloseButton onClose={() => setAlert({ variant: 'success', show: false, title: ''})} />} variant={alert.variant} title={alert.title} />}
                <CardBody>
                    <Form id="enable-user-form">
                    <FormGroup label="Username" fieldId="enable-username" isRequired>
                        <TextInput
                        isRequired
                        type="text"
                        id="enable-username"
                        name="enable-username"
                        value={username}
                        onChange={(value) => setUsername(value)}
                        placeholder="Username"
                        />
                    </FormGroup>

                    <ActionGroup>
                        <Button id="enable-user-submit" variant="primary" onClick={promptEnableUser}>Enable User</Button>
                    </ActionGroup>
                    </Form>
                </CardBody>
            </Card>
        </PageSection>
    </>
    );
}
