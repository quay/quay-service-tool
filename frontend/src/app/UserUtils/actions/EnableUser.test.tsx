import { mount, shallow } from 'enzyme';
import React from 'react';
import axios from "axios";
import { EnableUser } from './EnableUser';
import { List, Modal } from '@patternfly/react-core';
import HttpService from "../../../services/HttpService";
import { mocked } from 'ts-jest/utils';
import {act} from 'react-dom/test-utils';

jest.mock('../../../services/HttpService', () => ({
    axiosClient: {
        get: jest.fn(),
        put: jest.fn()
    }
}));

// Wait for all pending promises to resolve
// Adds promise to end of event loop and waits for it to resolve
function flushPromises() {
    return new Promise(resolve => setImmediate(resolve));
}

async function typeAndSubmitUsername(view, username){
    let input = view.find("input#enable-username").at(0);
    await act(async () => {
        input.getDOMNode().setAttribute('value', username);
        input.simulate('change', {currentTarget: input});
    });
    await act(async () => {
        view.find('button#enable-user-submit').simulate('click');
        await flushPromises(); // Pauses execution of tests until promises in the component have been resolved
        view.update(); // Sync enzyme component tree with the react component tree
    });
}

describe('Enable users tests', ()=>{
    it('should render', () => {
        const view = mount(<EnableUser />);
        expect(view).toMatchSnapshot();
    });

    it('Should alert error if username is not provided', () => {
        const view = mount(<EnableUser />);
        view.find('button#enable-user-submit').simulate('click');
        expect(view.find('Alert#enable-user-alert').props()).toHaveProperty('variant', 'danger');
        expect(view.find('Alert#enable-user-alert').text().includes('Please enter a username')).toBe(true);
    })

    it('Should alert if given username isn\'t found in backend', async () => {
        let username = 'nonexistentusername';
        mocked(HttpService, true).axiosClient.get.mockRejectedValue({response: {status: 404, data:{message: `Could not find user ${username}`}}});
        const view = mount(<EnableUser />);
        await typeAndSubmitUsername(view, username);
        expect(view.find('Alert#enable-user-alert').props()).toHaveProperty('variant', 'danger');
        expect(view.find('Alert#enable-user-alert').text().includes(`User ${username} does not exist`)).toBe(true);
    })

    it('Should alert if user is already enabled', async ()=>{
        mocked(HttpService, true).axiosClient.get.mockResolvedValue({data: {enabled: true}});
        let username = "nonexistentusername";
        const view = mount(<EnableUser />);
        await typeAndSubmitUsername(view, username);
        expect(view.find('Alert#enable-user-alert').props()).toHaveProperty('variant', 'danger');
        expect(view.find('Alert#enable-user-alert').text().includes(`User ${username} is already enabled`)).toBe(true);
    })

    it('Should enable user', async()=>{
        let username = "existinguser";
        mocked(HttpService, true).axiosClient.get.mockResolvedValue({data: {user: username, enabled: false}});
        mocked(HttpService, true).axiosClient.put.mockResolvedValue({data: {message: "User updated successfully", user: "existinguser",enabled: true}});
        const view = mount(<EnableUser />);
        await typeAndSubmitUsername(view, username);
        expect(view.find('Modal').props()).toHaveProperty('isOpen', true);
        expect(view.find('Modal').props()).toHaveProperty('title', `Enable user ${username}?`);
        await act(async () => {
            view.find('button#enable-user-confirm').simulate('click');
            await flushPromises(); 
            view.update();
        });
        expect(view.find('Alert#enable-user-alert').props()).toHaveProperty('variant', 'success');
        expect(view.find('Alert#enable-user-alert').text().includes(`User ${username} enabled`)).toBe(true);
        expect(view.find('Modal').exists()).toBe(false);
    })
});
