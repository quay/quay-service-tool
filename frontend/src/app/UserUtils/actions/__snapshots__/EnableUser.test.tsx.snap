// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Enable users tests should render 1`] = `
<EnableUser>
  <PageSection>
    <section
      className="pf-c-page__main-section"
    >
      <Card>
        <article
          className="pf-c-card"
          data-ouia-component-id="OUIA-Generated-Card-1"
          data-ouia-component-type="PF4/Card"
          data-ouia-safe={true}
          id=""
        >
          <CardTitle
            className="text-uppercase"
          >
            <div
              className="pf-c-card__title text-uppercase"
            >
               Enable User 
            </div>
          </CardTitle>
          <CardBody>
            <div
              className="pf-c-card__body"
            >
              <Form
                id="enable-user-form"
              >
                <form
                  className="pf-c-form"
                  id="enable-user-form"
                  noValidate={true}
                >
                  <FormGroup
                    fieldId="enable-username"
                    isRequired={true}
                    label="Username"
                  >
                    <div
                      className="pf-c-form__group"
                    >
                      <div
                        className="pf-c-form__group-label"
                      >
                        <label
                          className="pf-c-form__label"
                          htmlFor="enable-username"
                        >
                          <span
                            className="pf-c-form__label-text"
                          >
                            Username
                          </span>
                          <span
                            aria-hidden="true"
                            className="pf-c-form__label-required"
                          >
                             
                            *
                          </span>
                        </label>
                         
                      </div>
                      <div
                        className="pf-c-form__group-control"
                      >
                        <TextInput
                          id="enable-username"
                          isRequired={true}
                          name="enable-username"
                          onChange={[Function]}
                          placeholder="Username"
                          type="text"
                          value=""
                        >
                          <TextInputBase
                            aria-label={null}
                            className=""
                            id="enable-username"
                            innerRef={null}
                            isDisabled={false}
                            isLeftTruncated={false}
                            isReadOnly={false}
                            isRequired={true}
                            name="enable-username"
                            onChange={[Function]}
                            ouiaSafe={true}
                            placeholder="Username"
                            type="text"
                            validated="default"
                            value=""
                          >
                            <input
                              aria-invalid={false}
                              aria-label={null}
                              className="pf-c-form-control"
                              data-ouia-component-id="OUIA-Generated-TextInputBase-1"
                              data-ouia-component-type="PF4/TextInput"
                              data-ouia-safe={true}
                              disabled={false}
                              id="enable-username"
                              name="enable-username"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              placeholder="Username"
                              readOnly={false}
                              required={true}
                              type="text"
                              value=""
                            />
                          </TextInputBase>
                        </TextInput>
                      </div>
                    </div>
                  </FormGroup>
                  <ActionGroup>
                    <div
                      className="pf-c-form__group pf-m-action"
                    >
                      <div
                        className="pf-c-form__group-control"
                      >
                        <div
                          className="pf-c-form__actions"
                        >
                          <Button
                            id="enable-user-submit"
                            onClick={[Function]}
                            variant="primary"
                          >
                            <ButtonBase
                              id="enable-user-submit"
                              innerRef={null}
                              onClick={[Function]}
                              variant="primary"
                            >
                              <button
                                aria-disabled={false}
                                aria-label={null}
                                className="pf-c-button pf-m-primary"
                                data-ouia-component-id="OUIA-Generated-Button-primary-1"
                                data-ouia-component-type="PF4/Button"
                                data-ouia-safe={true}
                                disabled={false}
                                id="enable-user-submit"
                                onClick={[Function]}
                                role={null}
                                type="button"
                              >
                                Enable User
                              </button>
                            </ButtonBase>
                          </Button>
                        </div>
                      </div>
                    </div>
                  </ActionGroup>
                </form>
              </Form>
            </div>
          </CardBody>
        </article>
      </Card>
    </section>
  </PageSection>
</EnableUser>
`;
